src/buf/fixed.rs

```diff
@@ -0,0 +1,184 @@
+//! Buffers pre-registered with the kernel.
+//!
+//! This module provides facilities for registering in-memory buffers with
+//! the `tokio-uring` runtime. Operations like [`File::read_fixed_at`][rfa] and
+//! [`File::write_fixed_at`][wfa] make use of buffers pre-mapped by
+//! the kernel to reduce per-I/O overhead.
+//! The [`BufRegistry::register`] method is used to register a collection of
+//! buffers with the kernel; it must be called before any of the [`FixedBuf`]
+//! handles to the collection's buffers can be used with I/O operations.
+//!
+//! [rfa]: crate::fs::File::read_fixed_at
+//! [wfa]: crate::fs::File::write_fixed_at
+
+use super::{IoBuf, IoBufMut};
+use crate::driver::{self, Buffers};
+
+use std::cell::RefCell;
+use std::io;
+use std::mem::ManuallyDrop;
+use std::rc::Rc;
+
+/// An indexed collection of I/O buffers pre-registered with the kernel.
+///
+/// `BufRegistry` allows the application to manage a collection of buffers
+/// allocated in memory, that can be registered in the current `tokio-uring`
+/// context using the [`register`] method.
+///
+/// [`register`]: Self::register
+///
+/// A `BufRegistry` value is a lightweight handle for a collection of
+/// allocated buffers. Cloning of a `BufRegistry` creates a new reference to
+/// the same collection of buffers.
+///
+/// The buffers of the collection are not deallocated until:
+/// - all `BufRegistry` references to the collection have been dropped;
+/// - all [`FixedBuf`] handles to individual buffers in the collection have
+///   been dropped, including the buffer handles owned by any I/O operations
+///   in flight;
+/// - every `tokio-uring` runtime the buffers have been registered in
+///   has terminated.
+#[derive(Clone)]
+pub struct BufRegistry {
+    inner: Rc<RefCell<Buffers>>,
+}
+
+/// A unique handle to a memory buffer that can be pre-registered with
+/// the kernel for `io-uring` operations.
+///
+/// `FixedBuf` handles can be obtained from a [`BufRegistry`] collection.
+/// For each buffer, only a single `FixedBuf` handle can be either used by the
+/// application code or owned by an I/O operation at any given time,
+/// thus avoiding data races between `io-uring` operations in flight and
+/// the application accessing buffer data.
+pub struct FixedBuf {
+    registry: Rc<RefCell<Buffers>>,
+    buf: ManuallyDrop<Vec<u8>>,
```

Comment:
`Box<[T]>`?

Reply:
@oliverbunting I don't think it's  possible to create a Box around a variable sized slice because trait Sized is not implemented for [u8]. Did you mean something else?

Reply:
A Vec also has the nice properties of len and cap, not just len, and I think this PR was taking advantage of an initial length, perhaps to avoid any initialization of the data being given to the kernel, or maybe to pre-seed the buffers with some values - perhaps to be used for a fixed buffer write. It's just the growable that I don't like but lots of crates seem to have decided to just not let their types grow by not calling things like push.

Reply:
@FrankReh I've long lamented that (as far as I know) there is no way of specifying that you _only_ want a generic `T` to be by value, and not by reference. `AsRef<[T]>` and `Deref` are just too permissive here

Reply:
@ollie-etl Your point is that the API this feature introduces doesn't really give ownership of the memory to the driver?

Reply:
No, it does. My point is I'd very much like to be able to generalize over any type T which can be `Deref`'d to a `[u8]`, but unfortunately I can't exclude `&[u8]` from that set, which prevents the generalization.

Reply:
Hmm, I have a suggestion - I'll do a full review and see if it sticks

Reply:
> Hmm, I have a suggestion - I'll do a full review and see if it sticks

I've looked into this, its much broader than the scope of this PR. This PR should not be blocked on my musings, it is useful as is, and can always be changed to a more general form in the future.

Reply:
`Vec` is used because we need to pass the original capacity on deallocation. Once the buffers are owned by the registry instance, they are not grown or shrunk.

The discussion made me think of adding an allocator type parameter so we can also accommodate `Vec` with custom allocators.